tt_metal/impl/device/mesh_device.hpp

```diff
@@ -4,28 +4,103 @@
 #include "tt_metal/impl/device/device.hpp"
 #include "tt_metal/impl/device/mesh_device_view.hpp"
 
 namespace tt::tt_metal {
 
 using DeviceIds = std::vector<int>;
+using MeshDeviceID = std::size_t;
 class MeshDeviceView;
 
-class MeshDevice
-{
-public:
-    MeshShape mesh_shape;
-    std::map<chip_id_t, Device *> managed_devices;
-    std::vector<std::pair<int, Device *>> mesh_devices;
-    std::shared_ptr<MeshDeviceView> view;
+// SystemMesh creates a virtualization over the physical devices in the system.
+// It creates a logical 2D-mesh of devices and manages the mapping between logical and physical device coordinates.
+// It is responsible for the assignment of devices in a MeshDevice to physical devices, and the creation and deletion of
+// device resources.
+class SystemMesh {
+    using LogicalCoordinate = Coordinate;
+    using PhysicalCoordinate = eth_coord_t;
+
+    // Keep track of the devices that were opened so we can close them later. We shouldn't
+    // to keep track of this but DevicePool seems to open all devices associated with an MMIO device id
+    std::unordered_map<MeshDeviceID, std::map<chip_id_t, Device*>> opened_devices;
+    std::unordered_map<MeshDeviceID, std::vector<chip_id_t>> assigned_devices;
+    std::unordered_map<MeshDeviceID, std::shared_ptr<MeshDevice>> assigned_mesh_device_devices;
+    std::unordered_map<chip_id_t, Device *> assigned_physical_id_to_device;
+
+    // Logical mesh shape and coordinates
+    MeshShape logical_mesh_shape;
+    std::map<LogicalCoordinate, PhysicalCoordinate> logical_to_physical_coordinates;
+
+    // Handling of physical coordinates
+    std::unordered_map<PhysicalCoordinate, chip_id_t> physical_coordinate_to_device_id;
+    std::unordered_map<chip_id_t, PhysicalCoordinate> physical_device_id_to_coordinate;
+
+    SystemMesh() = default;
+    SystemMesh(const SystemMesh &) = delete;
+    SystemMesh &operator=(const SystemMesh &) = delete;
+    SystemMesh(SystemMesh &&) = delete;
+    SystemMesh &operator=(SystemMesh &&) = delete;
+
+    static MeshShape get_system_mesh_shape(std::size_t system_num_devices);
+    static std::map<LogicalCoordinate, PhysicalCoordinate> get_system_mesh_translation_map(
+        std::size_t system_num_devices);
+
+    bool is_system_mesh_initialized() const;
+
+   public:
+    static SystemMesh &instance();
+
+    void initialize();
+
+    // Return the shape of the logical mesh
+    const MeshShape &get_shape() const;
+    std::size_t get_num_devices() const;
+
+    // Get the physical device IDs mapped to a MeshDevice
+    std::vector<chip_id_t> get_mapped_physical_device_ids(
+        const std::size_t requested_rows,
+        const std::size_t requested_cols,
+        const std::size_t row_offset = 0,
+        const std::size_t col_offset = 0) const;
+
+    // Map MeshDevice to physical devices
+    std::vector<Device *> map_mesh_device(
+        std::shared_ptr<MeshDevice> mesh_device,
+        size_t num_command_queues,
```

Comment:
might be good  to add small struct with 3 size_t params instead of 3 parameters of the same type.
Less change to make mistake.

Reply:
It's a good point but I decided to leave this alone for now and have this changed wholesale with @patrickroberts proposed changes in: https://github.com/tenstorrent/tt-metal/pull/12681